Source/monster.cpp

```diff
@@ -4912,83 +4915,50 @@ BOOL CheckNoSolid(int x, int y)
 
 BOOL LineClearF(BOOL (*Clear)(int, int), int x1, int y1, int x2, int y2)
 {
-	int xorg, yorg;
-	int dx, dy;
-	int d;
-	int xincD, yincD, dincD, dincH;
-	int tmp;
-	BOOL done = FALSE;
+	int dx = x2 - x1;
+	int dy = y2 - y1;
+	int steps;
+	BOOL done = false;
 
-	xorg = x1;
-	yorg = y1;
-	dx = x2 - x1;
-	dy = y2 - y1;
-	if (abs(dx) > abs(dy)) {
-		if (dx < 0) {
-			tmp = x1;
-			x1 = x2;
-			x2 = tmp;
-			tmp = y1;
-			y1 = y2;
-			y2 = tmp;
-			dx = -dx;
-			dy = -dy;
-		}
-		if (dy > 0) {
-			d = 2 * dy - dx;
-			dincD = 2 * dy;
-			dincH = 2 * (dy - dx);
-			yincD = 1;
-		} else {
-			d = 2 * dy + dx;
-			dincD = 2 * dy;
-			dincH = 2 * (dx + dy);
-			yincD = -1;
-		}
-		while (!done && (x1 != x2 || y1 != y2)) {
-			if ((d <= 0) ^ (yincD < 0)) {
-				d += dincD;
-			} else {
-				d += dincH;
-				y1 += yincD;
-			}
-			x1++;
-			done = ((x1 != xorg || y1 != yorg) && !Clear(x1, y1));
-		}
-	} else {
-		if (dy < 0) {
-			tmp = y1;
-			y1 = y2;
-			y2 = tmp;
-			tmp = x1;
-			x1 = x2;
-			x2 = tmp;
-			dy = -dy;
-			dx = -dx;
-		}
-		if (dx > 0) {
-			d = 2 * dx - dy;
-			dincD = 2 * dx;
-			dincH = 2 * (dx - dy);
-			xincD = 1;
+	if (abs(dx) > abs(dy))
+		steps = abs(dx);
+	else
+		steps = abs(dy);
+
+	if (steps == 0)
+		return false;
+
+	float Xinc = (float)dx / steps;

```

Comment:
Can we avoid floating point math for this?

Context: For the past 6 months I've been trying to make DevilutionX run well on a device without an FPU (RG99 jz4725).

https://en.wikipedia.org/wiki/Bresenham%27s_line_algorithm

Reply:
> Can we avoid floating point math for this?
> 
> Context: For the past 6 months I've been trying to make DevilutionX run well on a device without an FPU (RG99 jz4725).
> 
> https://en.wikipedia.org/wiki/Bresenham%27s_line_algorithm

Bresenham's line algorithm is what original Diablo uses and the the cause of one of the "4 tile" scenarios, specifically the one where monsters shoot into walls/corners endlessly. Bresenham's is a special case of DDA that only draws in one direction, but an arrow's path is NOT the same path in both directions between two points. Ajenbo and I discussed this over discord, so below are example lines drawn by Bresenham's and a proper line algorithm to highlight the difference.

Bresenham's drawing from origin (5, 1) to (0, 0) will draw it as {0x0, 1x0, 2x0, 3x1 4x1, 5x1}, but that's a line from (0, 0) to (5, 1). A line from (5, 1) to (0, 0) should actually be {5x1, 4x1, 3x1, 2x1, 1x1, 0x0} if I'm doing the DDA in this PR correctly in my head. Notice how (2, 1) is being checked in the PRs algorithm while it is not in Bresenham's. If an arrow dies on tile (2, 1) but not (2, 0), a monster standing at (5, 1) will think it can hit (2, 0) while it cannot. Same with the difference between (1, 1) and (1, 0).

I think the line it draws is also wrong in general, as Diablo's missile physics aim to have missiles "just barely" enter the target tile, i.e. they get hit the very "corner" of what would be considered "in the tile" so in a case like this, from (0, 0) to (5, 1) where one axis changes by 1 and the other by more, the axis that changes by 1 is going to change on the very last step. Bresenham's was changing that axis too early, as shown in the example above.

You could maybe do away with the float math and division with ints and bit shifting like other parts of Diablo does? Not sure how that might impact math with rounding though, but this already has issues with rounding into irrational decimals and not stepping when it should in the shorter axis.